package/gluon-mesh-vpn-wireguard/src/gluon-mesh-vpn-key-translate.c

```diff
@@ -0,0 +1,82 @@
+// SPDX-FileCopyrightText: 2022 Jan-Niklas Burfeind <gluon@aiyionpri.me>
+// SPDX-License-Identifier: BSD-2-Clause
+// SPDX-FileContributor: read_key() by Matthias Schiffer <mschiffer@universe-factory.net>
+
+#include <libubox/utils.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define FASTD_KEY_LEN_HEX 64
+
+static inline void invalid_key(void) {
+	fprintf(stderr, "Error: Invalid fastd key provided.\n");
+	exit(1);
+}
+
+static inline void no_mem(void) {
+	fprintf(stderr, "Error: Not enough memory for base64 encoding available.\n");
+	exit(1);
+}
+
+static inline void usage(void) {
+	fprintf(stderr, "Usage: gluon-mesh-vpn-key-translate [-n] <fastd secret>\n");
+	fprintf(stderr, "-n     do not output the trailing newline\n");
+	exit(1);
+}
+
+static inline bool read_key(uint8_t key[32], const char *hexkey) {
+	if ((strlen(hexkey) != 64) || (strspn(hexkey, "0123456789abcdefABCDEF") != 64))
+		return false;
+
+	size_t i;
+	for (i = 0; i < 32; i++)
+		sscanf(&hexkey[2 * i], "%02hhx", &key[i]);
+
+	return true;
+}
+
+int main(int argc, char *argv[]) {
+	bool newline = true;
+	if (argc < 2 || argc > 3) {
+		usage();
+	} else if (argc == 3) {
+		if (strcmp(argv[1], "-n")) {
+			usage();
+		} else {
+			newline = false;
+		}
+	}
+	int fastd_pos = argc - 1;
+	int fastd_len = strlen(argv[fastd_pos]);
+
+	if (FASTD_KEY_LEN_HEX != fastd_len)
+		invalid_key();
+
+	uint8_t fastd_keystore[32];
+	read_key(fastd_keystore, argv[fastd_pos]);
```

Comment:
You absolutely never want to pass secrets as command line arguments, as arguments are visible in procfs. Pass the key through stdin.

Reply:
Reading is now done via stdin only.